client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1902 @@
+
+	/// Indicate that a connection with the peer was dropped.
+	/// Returns `Err(PeerId)` if the peer wasn't connected or is not known to us.
+	fn on_peer_dropped_inner(&mut self, peer_id: PeerId) -> Result<(), PeerId> {
+		if self.drop_reserved_peer(&peer_id)? || self.drop_regular_peer(&peer_id) {
+			// The peer found and disconnected.
+			self.report_disconnect(peer_id);
+			Ok(())
+		} else {
+			// The peer was not found in neither regular or reserved lists.
+			Err(peer_id)
+		}
+	}
+
+	/// Try dropping the peer as a reserved peer. Return `Ok(true)` if the peer was found and
+	/// disconnected, `Ok(false)` if it wasn't found, `Err(PeerId)`, if the peer found, but not in
+	/// connected state.
+	fn drop_reserved_peer(&mut self, peer_id: &PeerId) -> Result<bool, PeerId> {
+		let Some(state) = self.reserved_nodes.get_mut(peer_id) else {
+			return Ok(false)
+		};
+
+		if let PeerState::Connected(direction) = state {
+			trace!(target: LOG_TARGET, "Reserved peer {peer_id} ({direction:?}) dropped.");
+			*state = PeerState::NotConnected;
+			Ok(true)
+		} else {
+			Err(*peer_id)
+		}
+	}
+
+	/// Try dropping the peer as a regular peer. Return `true` if the peer was found and
+	/// disconnected, `false` if it wasn't found.
+	fn drop_regular_peer(&mut self, peer_id: &PeerId) -> bool {
+		let Some(direction) = self.nodes.remove(peer_id) else {
+			return false
+		};
+
+		trace!(target: LOG_TARGET, "Peer {peer_id} ({direction:?}) dropped.");
+
+		match direction {
+			Direction::Inbound => self.num_in -= 1,
+			Direction::Outbound => self.num_out -= 1,
+		}
+
+		true
+	}
+
+	/// Initiate outgoing connections trying to connect all reserved nodes and fill in all outgoing
+	/// slots.
+	fn alloc_slots(&mut self) {
+		// Try connecting to reserved nodes first, ignoring nodes with outstanding events/actions.
+		self.reserved_nodes
+			.iter_mut()
+			.filter_map(|(peer_id, state)| {
+				(!state.is_connected() && !self.peer_store.is_banned(peer_id)).then(|| {
+					*state = PeerState::Connected(Direction::Outbound);
+					peer_id
+				})
+			})
+			.cloned()
+			.collect::<Vec<_>>()
+			.into_iter()
+			.for_each(|peer_id| {
+				self.start_connection(peer_id);
+			});
+
+		// Nothing more to do if we're in reserved-only mode or don't have slots available.
+		if self.reserved_only || self.num_out >= self.max_out {
+			return
+		}
+
+		// Fill available slots.
+		let available_slots = (self.max_out - self.num_out).saturated_into();
+
+		// Ignore reserved nodes (connected above), already connected nodes, and nodes with
+		// outstanding events/actions.
+		let ignored = self
+			.reserved_nodes
+			.keys()
+			.collect::<HashSet<&PeerId>>()
+			.union(&self.nodes.keys().collect::<HashSet<&PeerId>>())
+			.cloned()
+			.collect();
+
+		let candidates = self
+			.peer_store
+			.outgoing_candidates(available_slots, ignored)
+			.into_iter()
+			.filter_map(|peer_id| {
+				(!self.reserved_nodes.contains_key(&peer_id) && !self.nodes.contains_key(&peer_id))
```

Comment:
why are we checking `reserved_nodes` and `nodes` again if `ignored` already contains both sets?

Reply:
Because it's `PeerStore` responsibility to respect `ignored`, but it's a different component and if it doesn't behave up to spec, we must be able to detect it here.

Reply:
I would use a `debug_assert!` in case of a simpler condition, but it's not that convenient with filtering